### DISCUSSION TITLE
Pasting a path to a file/folder

### DISCUSSION THREAD

**Discussion by max-baz** [sayanarijit/xplr#245] [Ideas]
I use xplr as a file picker for the browser, so when I click "attach file" on some website, xplr will show up. Works nice, but I have a case when I already have a path to the file that I want to attach in clipboard. I wish to be able to press some button in xplr and just paste the path into xplr, so that xplr just goes into that dir and focuses on the file, so that I only need to press Enter to close xplr.

If i have a path to a directory, I guess xplr should enter the directory and focus on the first file there, or maybe just go to the parent directory and focus the directory I have in clipboard.

Do you think it's something that should live in my dotfiles, or is it something common enough to include in default config?

**Comment by sayanarijit** [comment]
First of all, nice! I didn't know xplr could do that 😅 
 
As I understand it requires gui interaction, and/or depends on clipboard tools. While there are exceptions like xdg-open, and editor, I'd avoid adding any more platform dependent tools.

For requirements like this, I'd first go with creating a plugin or a hack and see how useful the feature is, i.e. how many users use it. Only then we'll be able determine if it can be added to default as an exception like xdg-open.

**Reply by max-baz** [reply]
I was actually thinking of something simpler, like I press a button in xplr, an input prompt appears asking "enter path where you want to go", I press `ctrl+shift+v` to paste contents of my clipboard into the input prompt and hit enter 🙂

**Reply by sayanarijit** [reply]
Ah ok... I think that's possible. Maybe we can map it to `:o`.

**Reply by max-baz** [reply]
We can also start as a hack if you think it's too niche (but I'd ask for some help, how to implement this 😄 )

**Reply by sayanarijit** [reply]
Try this

```lua
local function setup()
  xplr.config.modes.builtin.action.key_bindings.on_key["o"] = {
    help = "visit file",
    messages = {
      "PopMode",
      { SwitchModeCustom = "visit_path" },
    },
  }

  xplr.config.modes.custom.visit_path = {
    name = "visit path",
    key_bindings = {
      on_key = {
        enter = {
          messages = {
            "FocusPathFromInput",
            "PopMode",
          }
        },
        esc = {
          help = "cancel",
          messages = {
            "PopMode"
          },
        },
        ["ctrl-c"] = {
          help = "terminate",
          messages = {
            "Terminate"
          }
        }
      },
      default = {
        messages = {
          "BufferInputFromKey",
        }
      }
    },
  }
end

return { setup = setup }
```

**Reply by max-baz** [reply]
Wow I didnt expect you to already have infrastructure around, like `FocusPathFromInput`! 😄 

It works with a few caveats that we might be able to address in the code:

- `~` is not accepted, should be auto-expanded to `/home/maximbaz`
- ~~When I get recovery warning (or some error), and then press `:o` again, the input prompt is not rendered, it's hidden behind the warning (but if I go to search using `/`, then it always shows the input prompt)~~ 
    - nevermind, discovered `{ SetInputBuffer = "" },`
    - (I also added shortcuts for backspace, ctrl+w, ctrl+u), nice
- If I put incorrect path, the error message is not optimal, but this is very minor

**Reply by sayanarijit** [reply]
We can replace `FocusPathFromInput` with `BashExecSilently = [===[FocusPath: echo $(ls -d "${XPLR_INPUT_BUFFER:?}") >> "${XPLR_PIPE_MSG_IN:?}"]===]`.

**Reply by sayanarijit** [reply]
`ls -d`, not `ls -ld`

**Reply by sayanarijit** [reply]
```
➜  ~ /bin/ls "~/Downloads"
/bin/ls: cannot access '~/Downloads': No such file or directory
```

Doesn't work. Not sure if adding shell expansion in rust will be a good idea. Because there are all kinds of shell dependent special characters.

**Reply by sayanarijit** [reply]
I don't have an optimal solution for this. Both Rust and Lua depend on external libraries to expand path. Maybe create an issue so that we can deal with it later.

**Reply by max-baz** [reply]
I'll try simply this hack, and see if it's sufficient for my case. Thanks for your help on this!

```lua
BashExecSilently = [===[
    dir="$(echo "${XPLR_INPUT_BUFFER:?}" | sed 's|~|/home/maximbaz|g')"
    echo "FocusPath: $dir" >> "${XPLR_PIPE_MSG_IN:?}"
]===],
```

**Comment by sayanarijit** [comment]
My mimeapps.list:

```mimeapps.list
[Added Associations]
# inode/directory=org.kde.dolphin.desktop;
inode/directory=xplr.desktop;
video/x-matroska=vlc.desktop;
x-scheme-handler/mailto=thunderbird.desktop;userapp-Evolution-3UUR00.desktop;
x-scheme-handler/tg=userapp-Telegram Desktop-S2PCX0.desktop;
image/png=firefox.desktop;

[Default Applications]
# inode/directory=org.kde.dolphin.desktop;
inode/directory=xplr.desktop;
video/x-matroska=vlc.desktop;
x-scheme-handler/http=firefox.desktop;
x-scheme-handler/https=firefox.desktop;
x-scheme-handler/mailto=userapp-Evolution-3UUR00.desktop
x-scheme-handler/tg=userapp-Telegram Desktop-S2PCX0.desktop
x-scheme-handler/etcher=appimagekit_baba58c0e81e6ce71aed830f21b8cca0-balenaEtcher.desktop
x-scheme-handler/graphql-playground=graphql-playground-electron.desktop
x-scheme-handler/web+stellar=keybase.desktop
x-scheme-handler/keybase=keybase.desktop
```

It works when I open a directory from let's say terminal, but I couldn't get firefox to use xplr for the picker.

**Reply by max-baz** [reply]
It's actually a configuration in the browser itself, e.g. in qutebrowser I configured it like so:

https://github.com/maximbaz/dotfiles/blob/df582b7d6a60b89487289006f7ba034bded4c5a0/.config/qutebrowser/config.py#L24-L25

mimeapps is for `xdg-open /some/dir`, as you have already discovered...

**Comment by sayanarijit** [comment]
@maximbaz You might want to check out https://github.com/sayanarijit/completion.xplr

Maybe I'll make a dedicated plugin. For now, it's a bit laggy, but I think it's usable.
